### DISCUSSION TITLE
Processing line sampling data

### DISCUSSION THREAD

**Discussion by joejoeyjoseph** [Exawind/amr-wind#451] [Q&A] [answered]
Hello, I am trying to generate and analyze the data vertically at a point, and I turned on the `LineSampler` option. After the simulation is finished, in `/post_processing/sampling0xxxx`, I notice different types of binary files exist (`DATA_0xxxx` and `Particle_H`). I am unsure which files contain the line sampling data, I cannot differentiate the two points/labels I specified (`ls1` and `ls2` below), and I do not know how to use the binary files efficiently. Please kindly give me some suggestions. 

The following is the setting in my abl file: 

`incflo.post_processing = sampling`
`sampling.output_frequency = 3000`
`sampling.output_format = "native"`
`sampling.labels = ls1 ls2`
`sampling.fields = velocity temperature`

`sampling.ls1.type = LineSampler`
`sampling.ls1.num_points = 200`
`sampling.ls1.start = 500.0 500.0 0.0`
`sampling.ls1.end = 500.0 500.0 1000.0`

`sampling.ls2.type = LineSampler`
`sampling.ls2.num_points = 200`
`sampling.ls2.start = 2800.0 2800.0 0.0`
`sampling.ls2.end = 2800.0 2800.0 1000.0`

Thank you in advance. 

Joseph

**Comment by sayerhs** [comment, marked as answer]
@joejoeyjoseph The data is saved in native AMReX particle format. You can use the [python script](https://github.com/Exawind/amr-wind/blob/main/tools/amrex_particle.py) in the `tools` directory to extract the data into a Pandas Dataframe. 

```python
from amrex_particle import AmrexParticleFile

# Load data from 1000th timestep in  the run directory (containing input file)
pfile = AmrexParticleFile.load(1000)

# get dataframe
df = pfile()
```

**Reply by joejoeyjoseph** [reply]
Thank you!

**Comment by joejoeyjoseph** [comment]
Hello, I tried to use the `AmrexParticleFile` function, and it gave me errors when I was loading the data. I figured out that the error is related to `pidx` and the offset between different sample towers in `amrex_particle`. I tried to debug it but I could not completely eliminate the error. 

I uploaded a directory of my sampling data at
https://github.com/joejoeyjoseph/awaken-code/tree/main/eagle/sampling60000 

Please let me know. Thank you in advance.

**Comment by sofYarikova** [comment]
Hi @sayerhs!
I am trying to turn on the LineSampler option and Averaging. 
The settings in my input file:

_incflo.post_processing = averaging line_sampling
averaging.type = TimeAveraging
averaging.labels = means
averaging.averaging_window = 10.0
averaging.averaging_start_time = 0
averaging.means.fields = velocity
averaging.means.averaging_type = ReAveraging
line_sampling.output_frequency = 10
line_sampling.labels = line0
line_sampling.fields = velocity_mean 
line_sampling.line0.type       = LineSampler
line_sampling.line0.num_points = 100
line_sampling.line0.start      = 0.1 0.0 -0.1
line_sampling.line0.end        = 0.1 0.0 0.1
line_sampling.output_format = ascii_


But after the simulation in finished files there are -nan velocity, i.e. I think that field was created, but nothing was determined there.  Why might this be so?

**Reply by sayerhs** [reply]
@sofYarikova I am tagging @tonyinme as he wrote the _time averaging_ code and has been using it for his work. 

I don't see anything obvious that could be wrong in the snippet you posted. I have a few things that I would suggest that we check to understand the root cause. If you have already done these steps, can you share your full input file so that we can recreate the error at our end. 

- The log file doesn't show any solution blow up in the equation residuals?
- The velocity field itself doesn't have any NaNs?

**Reply by tonyinme** [reply]
@sofYarikova , I agree with @sayerhs , please let us know if the solution itself if giving NaNs. If not, can you please share the input file?

**Comment by ews-ffarella** [comment]
Same issue here #1621

**Comment by ews-ffarella** [comment]
> Hi @sayerhs!
> 
> I am trying to turn on the LineSampler option and Averaging. 
> 
> The settings in my input file:
> 
> 
> 
> _incflo.post_processing = averaging line_sampling
> 
> averaging.type = TimeAveraging
> 
> averaging.labels = means
> 
> averaging.averaging_window = 10.0
> 
> averaging.averaging_start_time = 0
> 
> averaging.means.fields = velocity
> 
> averaging.means.averaging_type = ReAveraging
> 
> line_sampling.output_frequency = 10
> 
> line_sampling.labels = line0
> 
> line_sampling.fields = velocity_mean 
> 
> line_sampling.line0.type       = LineSampler
> 
> line_sampling.line0.num_points = 100
> 
> line_sampling.line0.start      = 0.1 0.0 -0.1
> 
> line_sampling.line0.end        = 0.1 0.0 0.1
> 
> line_sampling.output_format = ascii_
> 
> 
> 
> 
> 
> But after the simulation in finished files there are -nan velocity, i.e. I think that field was created, but nothing was determined there.  Why might this be so?

Hi! Did you ever solve this? I get the same

**Reply by ews-ffarella** [reply]
Solved in https://github.com/Exawind/amr-wind/issues/1621#issuecomment-2948093735
